(fix) RESTWS-972 translate nametemplate namemappings

xprl-gjf opened the pull request:
<!--- Add a pull request title above in this format -->
<!--- real example: 'RESTWS-738 Remove concept property setters from ObsResource classes' -->
<!--- 'RESTWS-JiraIssueNumber JiraIssueTitle' -->
## Description of what I changed
<!--- Describe your changes in detail -->
<!--- It can simply be your commit message, which you must have -->

The purpose of this change was:
1) apply locale-specific translations to the namemapping values when retrieving nametemplates (and consequently the dislplayText values of the `nametemplate.lines()` property, since these displayText values are derived from the namemapping values).
2) make the API implementation consistent for nametemplate and addresstemplate APIs

Changes applied:
1) Refactor `LayoutTemplateProvider` class by splitting out representation description methods into a separate `LayoutTemplateRepresentation` class.
2) Fix `LayoutTemplateProvider` to apply locale-specific translations to namemapping values.
3) Amend `AddressTemplateController2_0` to use a standard `BaseDelegatingConverter<AddressTemplate>` implementation when generating a REST response, as opposed to the custom hand-coded implementation that was previously used. 

## Issue I worked on
<!--- This project only accepts pull requests related to open issues -->
<!--- Want a new feature or change? Discuss it in an issue first -->
<!--- Found a bug? Point us to the issue/or create one so we can reproduce it -->
<!--- Just add the issue number at the end: -->
see https://issues.openmrs.org/browse/RESTWS-972

## Checklist: I completed these to help reviewers :)
<!--- Put an `x` in the box if you did the task -->
<!--- If you forgot a task please follow the instructions below -->
- [X] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [ ] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

_not applicable_

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [X] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [X] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [X] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

xprl-gjf commented:
@dkayiwa would you be available to review this?

xprl-gjf commented:
I performed integration testing of addresstemplate via GET /openmrs/ws/rest/v1/addresstemplate/

Both before and after this PR, the result was identical:
```
{
  "displayName": null,
  "codeName": "default",
  "country": null,
  "lines": [
    [
      {
        "isToken": "IS_NOT_ADDR_TOKEN",
        "displayText": ""
      },
      {
        "isToken": "IS_ADDR_TOKEN",
        "displayText": "Address",
        "codeName": "address1",
        "displaySize": "40"
      }
    ],
    [
      {
        "isToken": "IS_NOT_ADDR_TOKEN",
        "displayText": ""
      },
      {
        "isToken": "IS_ADDR_TOKEN",
        "displayText": "Neighborhood",
        "codeName": "address2",
        "displaySize": "40"
      }
    ],
    [
      {
        "isToken": "IS_NOT_ADDR_TOKEN",
        "displayText": ""
      },
      {
        "isToken": "IS_ADDR_TOKEN",
        "displayText": "City/Village",
        "codeName": "cityVillage",
        "displaySize": "60"
      }
    ],
    [
      {
        "isToken": "IS_NOT_ADDR_TOKEN",
        "displayText": ""
      },
      {
        "isToken": "IS_ADDR_TOKEN",
        "displayText": "District",
        "codeName": "countyDistrict",
        "displaySize": "30"
      },
      {
        "isToken": "IS_NOT_ADDR_TOKEN",
        "displayText": " "
      },
      {
        "isToken": "IS_ADDR_TOKEN",
        "displayText": "Province",
        "codeName": "stateProvince",
        "displaySize": "30"
      },
      {
        "isToken": "IS_NOT_ADDR_TOKEN",
        "displayText": " "
      },
      {
        "isToken": "IS_ADDR_TOKEN",
        "displayText": "Country",
        "codeName": "country",
        "displaySize": "12"
      }
    ]
  ],
  "lineByLineFormat": [
    "address1",
    "address2",
    "cityVillage",
    "countyDistrict stateProvince country"
  ],
  "nameMappings": {
    "country": "Country",
    "stateProvince": "Province",
    "countyDistrict": "District",
    "cityVillage": "City/Village",
    "postalCode": "Location.postalCode",
    "address2": "Neighborhood",
    "address1": "Address"
  },
  "sizeMappings": {
    "country": "12",
    "stateProvince": "30",
    "countyDistrict": "30",
    "cityVillage": "60",
    "postalCode": "16",
    "address2": "40",
    "address1": "40"
  },
  "elementDefaults": {
    "country": "Ecuador"
  },
  "elementRegex": null,
  "elementRegexFormats": null,
  "requiredElements": null
}
```

xprl-gjf commented:
I performed integration testing of GET /openmrs/ws/rest/v1/nametemplate/short/

Before:
```
{
  "displayName": "Short Name Format",
  "codeName": "short",
  "country": null,
  "lines": [
    [
      {
        "isToken": "IS_NOT_NAME_TOKEN",
        "displayText": ""
      },
      {
        "isToken": "IS_NAME_TOKEN",
        "displayText": "PersonName.givenName",
        "codeName": "givenName",
        "displaySize": "30"
      }
    ],
    [
      {
        "isToken": "IS_NOT_NAME_TOKEN",
        "displayText": ""
      },
      {
        "isToken": "IS_NAME_TOKEN",
        "displayText": "PersonName.middleName",
        "codeName": "middleName",
        "displaySize": "30"
      }
    ],
    [
      {
        "isToken": "IS_NOT_NAME_TOKEN",
        "displayText": ""
      },
      {
        "isToken": "IS_NAME_TOKEN",
        "displayText": "PersonName.familyName",
        "codeName": "familyName",
        "displaySize": "30"
      }
    ]
  ],
  "lineByLineFormat": [
    "givenName",
    "middleName",
    "familyName"
  ],
  "nameMappings": {
    "givenName": "PersonName.givenName",
    "familyName": "PersonName.familyName",
    "middleName": "PersonName.middleName"
  },
  "sizeMappings": {
    "givenName": "30",
    "familyName": "30",
    "middleName": "30"
  },
  "elementDefaults": null,
  "elementRegex": null,
  "elementRegexFormats": null,
  "requiredElements": null,
  "resourceVersion": "2.0"
}
```

After:
```
{
  "displayName": "Short Name Format",
  "codeName": "short",
  "country": null,
  "lines": [
    [
      {
        "isToken": "IS_NOT_NAME_TOKEN",
        "displayText": ""
      },
      {
        "isToken": "IS_NAME_TOKEN",
        "displayText": "Nombre",
        "codeName": "givenName",
        "displaySize": "30"
      }
    ],
    [
      {
        "isToken": "IS_NOT_NAME_TOKEN",
        "displayText": ""
      },
      {
        "isToken": "IS_NAME_TOKEN",
        "displayText": "Segundo Nombre",
        "codeName": "middleName",
        "displaySize": "30"
      }
    ],
    [
      {
        "isToken": "IS_NOT_NAME_TOKEN",
        "displayText": ""
      },
      {
        "isToken": "IS_NAME_TOKEN",
        "displayText": "Apellido",
        "codeName": "familyName",
        "displaySize": "30"
      }
    ]
  ],
  "lineByLineFormat": [
    "givenName",
    "middleName",
    "familyName"
  ],
  "nameMappings": {
    "middleName": "Segundo Nombre",
    "givenName": "Nombre",
    "familyName": "Apellido"
  },
  "sizeMappings": {
    "givenName": "30",
    "familyName": "30",
    "middleName": "30"
  },
  "elementDefaults": null,
  "elementRegex": null,
  "elementRegexFormats": null,
  "requiredElements": null,
  "resourceVersion": "2.0"
}
```

dkayiwa commented on omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/helper/LayoutTemplateProvider.java:
```diff
@@ -48,86 +38,79 @@ public LayoutTemplateProvider(LayoutSupport<T> layoutSupport, String layoutDefau
-		obj.add("requiredElements", instance.getRequiredElements());
-		return obj;
+	private T translateNameMappings(T template, Locale locale) {
+		MessageSourceService messages = Context.getMessageSourceService();
```
Is there a reason why you are calling this `messages` instead of something like `messageService`

xprl-gjf replied:
Renamed.

dkayiwa commented:
I changed my local to `français` or `fr` but the `displayText` was still in English instead of french. Did you test that too?

xprl-gjf commented:
> I changed my locale to `français` or `fr` but the `displayText`, as returned from the REST call, was still in English instead of french. Did you test that too?

Oops. Good catch. Upon further investigation I found that `LayoutTemplateProvider<T>` was modifying the (singleton instance) LayoutTemplate being returned by the encapsulated `LayoutSupport<T>` service!! This meant that the translated messages from the first call to the API were 'installed' in the template, and so no further translation was performed on subsequent calls.

Fixed by amending `LayoutTemplateProvider<T>` to make a shallow copy of the returned template before modifying it to apply translations.

Tested for both AddressTemplates and NameTemplates.

xprl-gjf commented:
@dkayiwa ready for re-review if you are available. Thanks.
